eva-accession-release-automation/gather_clustering_counts/gather_release_counts.py

```diff
@@ -0,0 +1,300 @@
+    if linked_set1 != source_linked_set1 or linked_set2 != source_linked_set2:
+        tmp_linked_set1, tmp_linked_set2 = find_link(linked_set1, dict1, dict2, linked_set1, linked_set2)
+        linked_set1.update(tmp_linked_set1)
+        linked_set2.update(tmp_linked_set2)
+    return frozenset(linked_set1), frozenset(linked_set2)
+
+
+def gather_assemblies_and_species_from_directory(release_directory):
+    """
+    This function creates two dictionaries by walking through the release directory:
+     - all_assemblies_2_species: keys are the assemblies found in the release directory and the values are the list of
+       species directory names associated with them.
+     - all_species_2_assemblies: keys are species directory names and values are the list of assembly they contains.
+       It also contains the list of species with unmapped data and in the case the assembly list is empty.
+    """
+    all_assemblies_2_species = defaultdict(list)
+    all_species_2_assemblies = defaultdict(list)
+    for assembly_dir in glob.glob(os.path.join(release_directory, '*', "GCA_*")):
+        assembly_accession = os.path.basename(assembly_dir)
+        species_name = os.path.basename(os.path.dirname(assembly_dir))
+        all_assemblies_2_species[assembly_accession].append(species_name)
+        all_species_2_assemblies[species_name].append(assembly_accession)
+    for unmapped_file in glob.glob(os.path.join(release_directory, '*', "*unmapped_ids.txt.gz")):
+        species_name = os.path.basename(os.path.dirname(unmapped_file))
+        if species_name not in all_species_2_assemblies:
+            all_species_2_assemblies[species_name] = []
+    return all_assemblies_2_species, all_species_2_assemblies
+
+
+def gather_count_for_set_species(release_directory, set_of_species, output_dir):
+    """This function find al the release files for a set of species and pass them to the counting bash script."""
+    all_files = collect_files_to_count(release_directory, set_of_species)
+    script_name = 'count_rs_for_all_files.sh'
+    input_file_list = os.path.join(output_dir, f'input_for_{"_".join(sorted(set_of_species))}.list')
+    with open(input_file_list, 'w') as open_file:
+        open_file.write("\n".join(all_files))
+    output_file = os.path.join(output_dir, f'count_for_{"_".join(sorted(set_of_species))}.log')
+    if not os.path.exists(output_file):
+        run_command_with_output(
+            f'Run {script_name} for {", ".join(set_of_species)}',
+            f'{os.path.join(shell_script_dir, script_name)} {output_file} {input_file_list} '
+        )
+    else:
+        logger.warning(f'output {output_file} already exists. Remove it to perform the count again')
+    return output_file
+
+
+def collect_files_to_count(release_directory, set_of_species):
+    all_files = []
+    for species in set_of_species:
+        species_dir = os.path.join(release_directory, species)
+        assembly_directories = glob.glob(os.path.join(species_dir, "GCA_*"))
```

Comment:
**Optional:** I am wondering if we shoud use the release tracker table as the source of truth instead of the release directory... This way we can also validate if all the intended species were released and update that table if not.